### ISSUE TITLE
WyRng state

### ISSUE THREAD

**Issue by palimondo** [cocowalla/wyhash-dotnet#1]
I'm just looking over the code without running this, but I'm curious about the `WyRng` implementation… should the state stored in `seed` be [also modified also by the `mum` operation](https://github.com/cocowalla/wyhash-dotnet/blob/master/src/WyHash/WyRng.cs#L31)? 

Looking at the original [wyrand](https://github.com/wangyi-fudan/wyhash/blob/master/wyhash.h#L76) implementation, I would expect seed to be modified only once by the prime addition on the first line as the results of mum is just returned directly. Am I mistaken?

**Comment by cocowalla** [comment]
The assumption I made is that each generated value is the seed for the next generation - but looking again, I'm pretty sure you're correct :anguished:!

I also had a look at @eldruin's Rust implementation, and it looks like he is doing the same as me, feeding the last generated value back into `wyrng` as the seed for the next.

@wangyi-fudan, would be much appreciated it if you could confirm the correct behaviour here, and/or ideally provide a test vector? (e.g. a starting seed, and then the next 10 values that would be generated from that seed).

**Comment by eldruin** [comment]
I wrote [some back-to-back tests](https://github.com/eldruin/wyhash-rs/blob/master/tests/back_to_back_prng.rs) that ensure that my implementation returns the same as when [using the original wyhash](https://github.com/eldruin/wyhash-rs/blob/master/original/main.cpp) so they are definitely consistent. That was important to me.
Whether the seed being modified by mum is more mathematically sound or not, would be a question for @wangyi-fudan.

**Comment by cocowalla** [comment]
@eldruin to be clear, I'm not suggesting that your implementation deviates from WangYi's - it doesn't :)

I also get the same results as your RNG implementation (I checked with your values, thanks!) - but that's because of how we're *using* the implementation in our tests.

If you look [here](https://github.com/eldruin/wyhash-rs/blob/master/original/main.cpp#L35) you can see that you are feeding the generated number back into `wyrng` as the seed, same as I do. WangYi's code doesn't demonstrate generating a sequence of values, but as it mutates the `seed` variable, it does seem to [*suggest*](https://github.com/wangyi-fudan/wyhash/blob/master/wyhash.h#L76) that it's not actually the generated values that are fed back in, but the mutated seed.

**Comment by eldruin** [comment]
@cocowalla Ok I understood the problem now, thanks :)
Then I would rephrase the question in code:
@wangyi-fudan which of the two versions below is the intended way of using `wyrng` to generate a sequence of values?
```c++
unsigned long long current = 0;
for (unsigned int i = 0; i < 10; ++i)
{
    current = wyrng(&current);
    printf("%x", current);
}
```
or 
```c++
unsigned long long seed = 0;
for (unsigned int i = 0; i < 10; ++i)
{
    unsigned long long current = wyrng(&seed);
    printf("%x", current);
}
```

**Comment by palimondo** [comment]
For context, I stumbled on this question updating the Swift port of the older version of that generator:
https://github.com/lemire/SwiftWyhash/issues/6#issuecomment-497285092

@lemire verified that the older version passes BigCrunch and it [modified the seed just once by addition of the first prime](https://github.com/lemire/testingRNG/blob/master/source/wyhash.h#L11) (i.e. the implementation as I understood in my question here). I haven't verified the current `wyrand` implementation, but here's a PR for that: https://github.com/lemire/testingRNG/pull/19

**Comment by wangyi-fudan** [comment]
unsigned long long seed = 0;
for (unsigned int i = 0; i < 10; ++i)
{
    unsigned long long current = wyrng(&seed);
    printf("%x", current);
}

**Comment by cocowalla** [comment]
@wangyi-fudan thanks for confirming!

@palimondo good catch, I'll update as required and add test cases.

**Comment by cocowalla** [comment]
Fix is in master, and I'll push to NuGet some time in the next few days - thanks again to @palimondo for catching this, and to @wangyi-fudan for confirming the correct behaviour!

**Comment by eldruin** [comment]
Thank you everybody from me for noticing, letting me know and confirming the correct behavior! I will fix this in the Rust implementation as well. :)

**Comment by cocowalla** [comment]
Just a quick note to say I finally got round to pushing these changes to [nuget](https://www.nuget.org/packages/WyHash/1.0.3)
